### PR TITLE
sql: add sql.mutations.mutation_batch_byte_size setting

### LINKED ISSUE #67542: sql: writing many non-small rows causes `command is too large`
Currently we chunk mutations in to batches that each process up to 10k input rows. This however can easily hit the KV command size limit of 64mb if those rows produce larger batch entries, e.g. with just 1mb rows:
```
demo@127.0.0.1:26257/movr> insert into src select generate_series(0, 300), repeat('0123456789abcdef', 65536);
ERROR: command is too large: 315630116 bytes (max: 67108864)
```

We should instead chunk into smaller batches based on the size of the produced KV batch, e.g. aiming to produce 4MB batches, controlled by a cluster setting. Obviously any single row must still be <= the limit, but right now we effectively require 10k+ row inserts to have a per-row size of no more than 6.5kb.

### PR DESCRIPTION
**Pull request by dt**
Previously we always constructed 10k row insert batches, regardless of the
size of those rows. With large rows, this could easily exceed the kv size
limit of 64MB. This changes batch construction to track the size of added
keys and values, and send the batch either when it has 10k entries of when
the size of added keys and values exceeds the setting, which defaults to 4MB.

Fixes #67542.

Release note (bug fix): INSERT and UPDATE statements which operate on larger rows are split into batches using the sql.mutations.mutation_batch_byte_size setting

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Nice, thanks!

In terms of testing, I think we usually use tracing to confirm the KV level operations. I'd probably look at `execbuilder/testdata/show_trace_nonmetamorphic` and `execbuilder/testdata/autocommit_nonmetamorphic` for inspiration.

> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-:-MeVnSi4Dl09xmMFFywH:bx38pff)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @nvanbenschoten)

___
*[pkg/sql/tablewriter.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-MeVmdl_1p9TDAC8XUzH:-MeVmdl_1p9TDAC8XUzI:b-7kggu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4324a18d4ef68a60f805178b2e4a8a6f29bc48e8/pkg/sql/tablewriter.go#L115)):*
> ```Go
> 	// a different value in tests.
> 	maxBatchSize int
> 	// maxBatchSize determines the maximum number of key and value bytes in the KV
> ```

nit: `s/maxBatchSize/maxBatchByteSize/`.

**Comment by dt** [comment]
It seems like there are still some cases where the eval'ed `WriteBatch` can explode to be much larger than the size of the arguments in the `kv.Batch` the client sends and we might still hit the raft limit, but I still think this is better than nothing: If we try to stick to sending 4MiB of args, that should at least make it less likely we'll hit the 64MiB write-batch limit than when we were sending 128MiB of args before. There just might be room to do even better w.r.t predicting the size of the writebatch, I'm not sure.

**Review by dt — COMMENTED** [review]
Thanks, the trace tests indeed look like they do what I want. Updated with a test case

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-:-MeWrUYV30_4YYWMMod2:b-hbk0xi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @yuzefovich)

**Review by RaduBerinde — COMMENTED** [review]
There is also the insert fast path (sql/insert_fast_path.go) which doesn't do any batching. It is used when we insert from a VALUES with less than MaxBatchSize rows:

https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/exec/execbuilder/mutation.go#L150

We'd need to change it to also do batching internally. (Or maybe check the values and only allow the fast path if they are not big).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-:-MeWsobq5s_EgmZ8rLl3:bpwmfwi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @yuzefovich)

**Comment by dt** [comment]
IMO a large `VALUES` clause _should_ batch internally as well but I'm less worried about it since the user can just send fewer thing, whereas our internal batch of *10,000* things from a non-VALUES plan is impossible to avoid? I'm not saying that we shouldn't _also_ batch a big VALUES clause, but I think even as is, this is better than nothing (and since this is fixing a bug that can blocking some RESTOREs, I was trying to minimize the behavior change to keep it backport-safe if I can)

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 6 of 15 files at r2, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/opt/exec/execbuilder/testdata/show_trace_nonmetamorphic, line 400 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Me_fEFK4heqdgiLqklH:-Me_fEFK4heqdgiLqklI:bamv816) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/sql/opt/exec/execbuilder/testdata/show_trace_nonmetamorphic#L400)):*
> ```
> query TT
> SELECT operation, message FROM [SHOW KV TRACE FOR SESSION]
> WHERE message     LIKE '%r$rangeid: sending batch%'
> ```

Hm, I'm a bit confused - I would think that new `blobs` table would have the next range id and we needed to update `$rangeid` logic test variable, do you know why that's not the case?

**Comment by dt** [comment]
```
                    testdata/show_trace_nonmetamorphic:415: SELECT 'a', range_id FROM [SHOW RANGES FROM TABLE a] UNION SELECT 'blobs', range_id FROM [SHOW RANGES FROM TABLE blobs]
                    expected:
                        a 40
                        blobs 41
                    but found (query options: "") :
                        a      40
                        blobs  40
```

**Review by yuzefovich — APPROVED** [review]
Oh well, nvm then. :lgtm: but might be worth for Nathan to take a quick look at `pkg/kv` change.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-:-Mea9-sJEDKG8m-wsQSo:b2w01km)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @yuzefovich)

**Comment by dt** [comment]
@nvanbenschoten I think in offline convo you indicated you were okay with the idea of adding some tracking to the kv client `Batch` to count the key/value bytes of puts/dels/etc added to the batch, but can you take a quick look at the implementation of that here to make sure it lines up with what you had in mind?

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 6 of 15 files at r2, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-:-Mf4tjb4BsnGt84pWGeV:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @dt and @yuzefovich)

___
*[pkg/kv/batch.go, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4sHEk-3X4vePmGuqm:-Mf4sHEk-3X4vePmGuqn:bospbjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/kv/batch.go#L56)):*
> ```Go
> 	AdmissionHeader  roachpb.AdmissionHeader
> 	reqs             []roachpb.RequestUnion
> 	mutationReqBytes int
> ```

Let's name this `approxMutationReqBytes` and give it a comment similar to what you have below.
___
*[pkg/kv/batch.go, line 495 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4s_eR0V-iqVNRD08P:-Mf4s_eR0V-iqVNRD08Q:b3tv28n) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/kv/batch.go#L495)):*
> ```Go
> 		b.appendReqs(roachpb.NewConditionalPut(k, v, expValue, allowNotExist))
> 	}
> 	b.mutationReqBytes += len(k) + len(expValue) + len(v.RawBytes)
> ```

I don't think we need to include `expValue` in this estimate. We won't be writing the `expValue`.
___
*[pkg/sql/tablewriter.go, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4tDpd2jXxioYuwN90:-Mf4tDpe2Lqv_wXFGLDS:b90nwuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/sql/tablewriter.go#L132)):*
> ```Go
> }
> 
> var maxBatchBytes = settings.RegisterByteSizeSetting(
> ```

Should this be pushed into the `mutations` package? It's strange to me that the setting lives here but the metamorphic const lives elsewhere.

Also, if we were going to revive the cluster setting for `MaxBatchSize`, where should that live. Maybe a question for @yuzefovich.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @dt, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/tablewriter.go, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4tDpd2jXxioYuwN90:-Mf58_oOAOtit1qlj4lg:b-rchksy) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/sql/tablewriter.go#L132)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this be pushed into the `mutations` package? It's strange to me that the setting lives here but the metamorphic const lives elsewhere.

Also, if we were going to revive the cluster setting for `MaxBatchSize`, where should that live. Maybe a question for @yuzefovich.
</blockquote></details>

Personally, I don't have a strong preference about the placement of these things, either seems reasonable to me.

The original `sql.testing.mutations.max_batch_size` setting lived in `mutations` package. Back then we also performed the "resolution" of the value in `FlowBase.Setup` which was updating a global value, and I actually prefer David's approach in `tableWriterBase.init`.

**Comment by adityamaru** [comment]
@dt how far back do we think this is back portable (if at all)? Asking because the customer that prompted this fix is looking to understand what we are doing to fix the issue.

**Comment by dt** [comment]
I was vaguely thinking I'd _like_ to backport to 20.2, but I'll defer to SQL-Queries TLs on backport call and if so, how far back

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @nvanbenschoten and @yuzefovich)

___
*[pkg/kv/batch.go, line 495 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4s_eR0V-iqVNRD08P:-MfAAMA68MFhWxOHPIxq:b-evuqku) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/kv/batch.go#L495)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't think we need to include `expValue` in this estimate. We won't be writing the `expValue`.
</blockquote></details>

Ah, right, I guess with prop-eval'ed KV we don't put the actual cput in the replicated batch, just the result
___
*[pkg/sql/tablewriter.go, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4tDpd2jXxioYuwN90:-MfAA_Ug9qL1W7nttXuw:bk2rm4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/sql/tablewriter.go#L132)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Personally, I don't have a strong preference about the placement of these things, either seems reasonable to me.

The original `sql.testing.mutations.max_batch_size` setting lived in `mutations` package. Back then we also performed the "resolution" of the value in `FlowBase.Setup` which was updating a global value, and I actually prefer David's approach in `tableWriterBase.init`.
</blockquote></details>

So leave it as-is?

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67537)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @dt, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/tablewriter.go, line 132 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67537#-Mf4tDpd2jXxioYuwN90:-MfAB3gmCeKcNMXtGsD3:b-u2jq2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/da02c25e798d3b8d7a6b1dc3c7916ddcd547ec1c/pkg/sql/tablewriter.go#L132)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

So leave it as-is?
</blockquote></details>

Yeah, sgtm.

**Comment by dt** [comment]
TFTRs!

bors r+

**Comment by dt** [comment]
@yuzefovich How do you feel about 21.1 and 20.2 backports here? This came up because `RESTORE` runs `INSERT INTO system.$tbl SELECT * FROM restored_systemdb.$tbl` and some customers were unable to RESTORE a table where we hit the command size limit, so I think it qualifies as a bug, not just an enhancement/feature but will defer to you.

**Comment by yuzefovich** [comment]
I think it is justifiable for a backport to both branches.

One thought there is whether we want to backport it in such a manner that by default the behavior doesn't change? I mean we would essentially set the default value of the new cluster setting to zero, and we'd be ignoring the setting unless it was changed to something greater than zero. This way we'd have a safe workaround for users who run into "command is too large" problems when restoring.

I guess I'm a bit worried about this change having some performance implications, and merging it to master seems ok to me given that we have plenty of time to - hopefully - catch perf regressions if they occur, but 2 weeks baking period might not be enough. E.g. if we want to insert 10k rows 5KB each, previously we would have a single batch with 50MB worth of mutations, and now we will have 13 batches - will 13 batches be slower?

One the other hand, though, I think we do recommend "pagination" of mutations on the app side, so maybe this example is an edge case and can be reasonably ignored.

**Comment by dt** [comment]
one wrinkle with requiring a cluster setting to opt into pagination to avoid hitting the limit and failing is that settings themselves are also restored during a RESTORE, so RESTORE pretty much always wants its `INSERT INTO x SELECT FROM y` to paginate

**Comment by yuzefovich** [comment]
I see. Ok, then I'll run some quick benchmarks to see whether my concern is valid.

**Comment by yuzefovich** [comment]
Hm, in my quick sanity check on the gceworker multiple smaller batches are faster for INSERT than a large single batch, so my concerns don't seem to be valid. I'll open up the backports and will merge after the 2 weeks baking period.

**Comment by yuzefovich** [comment]
@dt there are some complications with backporting this to 20.2 branch (we would probably need to cherry-pick some commits from #57483), is there a customer that needs this to be fixed on 20.2?

**Comment by dt** [comment]
@yuzefovich hm, yes, this was driven by a customer who has indeed asked for followup fixes and is on 20.2, but we could see if there is room to push back since it isn't clean. If we do want to fix on 20.2, maybe better to just add the setting and size limit in a more or less fresh change than try to backport with the #57483 changes?

**Comment by yuzefovich** [comment]
I agree, I'm currently doing some plumbing, and the backport isn't as clean, but the amount of additional changes is limited. My initial thought of having to cherry-pick things from #57483 was incorrect.

**Pull request merged**
